include/sleip/dynamic_array.hpp

```diff
@@ -33,250 +34,258 @@ struct dynamic_array
   using reverse_iterator       = std::reverse_iterator<iterator>;
   using const_reverse_iterator = std::reverse_iterator<const_iterator>;
 
+  static_assert(std::is_object_v<value_type>, "Only support object types");
+  static_assert(std::is_same_v<typename allocator_type::value_type, value_type>,
+                "Allocator's value type must match container's");
+
 private:
-  T*          data_ = nullptr;
-  std::size_t size_ = 0;
-  Allocator   alloc_;
+  struct deleter : boost::empty_value<Allocator, 0>
+  {
+    std::size_t size = 0;
+
+    deleter() noexcept(noexcept(Allocator()))
+      : boost::empty_value<Allocator, 0>(boost::empty_init_t{})
+    {
+    }
+
+    deleter(deleter const&) = delete;
+
+    deleter(deleter&& other) noexcept
+      : boost::empty_value<Allocator, 0>(
+          boost::empty_init_t{},
+          std::move(static_cast<boost::empty_value<Allocator, 0>&>(other).get()))
+      , size{other.size}
+    {
+      other.size = 0;
+    };
+
+    deleter(std::size_t size_)
+      : boost::empty_value<Allocator, 0>(boost::empty_init_t{}, Allocator())
+      , size{size_}
+    {
+    }
+
+    deleter(Allocator const& alloc, std::size_t size_)
+      : boost::empty_value<Allocator, 0>(boost::empty_init_t{}, alloc)
+      , size{size_}
+    {
+    }
+
+    auto
+    operator=(deleter const&) & -> deleter& = delete;
+
+    auto
+      operator=(deleter&& other) &
+      noexcept -> deleter&
+    {
+      boost::empty_value<Allocator, 0>::get() =
+        std::move(static_cast<boost::empty_value<Allocator, 0>&>(other).get());
+
+      size       = other.size;
+      other.size = 0;
+
+      return *this;
+    }
+
+    auto
+    get_allocator() const noexcept -> Allocator
+    {
+      return boost::empty_value<Allocator, 0>::get();
+    }
+
+    auto
+    operator()(T* ptr) -> void
+    {
+      auto alloc = get_allocator();
+      std::allocator_traits<Allocator>::deallocate(alloc, ptr, size);
+    }
+  };
+
+  std::unique_ptr<T, deleter> data_ = {nullptr, deleter(Allocator(), 0)};
 
 public:
-  dynamic_array() noexcept(noexcept(Allocator())) = default;
+  dynamic_array() noexcept(noexcept(Allocator())){};
 
   explicit dynamic_array(const Allocator& alloc) noexcept
-    : alloc_(alloc)
+    : data_(nullptr, deleter(alloc, 0))
   {
   }
 
   dynamic_array(size_type count, T const& value, Allocator const& alloc = Allocator())
-    : alloc_(alloc)
   {
-    data_ = std::allocator_traits<Allocator>::allocate(alloc_, count);
-
-#ifdef BOOST_NO_EXCEPTIONS
-    boost::alloc_construct_n(alloc_, data_, count, std::addressof(value), 1);
-    size_ = count;
-#else
+#ifndef BOOST_NO_EXCEPTIONS
     try {
-      boost::alloc_construct_n(alloc_, data_, count, std::addressof(value), 1);
-      size_ = count;
+#endif
+
+      auto alloc_ = alloc;
+      data_ = {std::allocator_traits<Allocator>::allocate(alloc_, count), deleter(alloc_, count)};
+      boost::alloc_construct_n(alloc_, data(), count, std::addressof(value), 1);
```

Comment:
Oh, I see the problem now, the same `unique_ptr` deleter is being used both for construction and for destruction.

What I had in mind was roughly:
```
unique_ptr<T[], dealloc> d = {std::allocator_traits<Allocator>::allocate(alloc_, count), dealloc{alloc_}};
boost::alloc_construct_n(alloc_, d_.get(), count, std::addressof(value), 1);
this->p = d.release();
```

Reply:
I gave a shot at implementing it this. It's actually quite clean, I think!